docs/madr/decisions/011-mesh-traffic-route.md

```diff
@@ -0,0 +1,964 @@
+   requestHeaderModifier:
+    set:
+     - name:
+       value:
+    add:
+     - name:
+       value:
+    remove:
+     - "name"
+   responseHeaderModifier:
+    set:
+     - name:
+       value:
+    add:
+     - name:
+       value:
+    remove:
+     - name
+   requestMirror:
+    backendRef:
+     kind: MeshService
+     name: backend-mirror
+   requestRedirect:
+    scheme:
+    hostname:
+    path:
+    port:
+    statusCode:
+   urlRewrite:
+    hostname:
+    path:
+     replaceFullPath:
+     replacePrefixMatch:
+   directResponse:
+    status:
+    body:
+     inlineBytes: <base64>
+     inlineString:
+     filename:
+     environmentVariable:
+```
+
+#### Routing
+
+If `matches` succeeds, the request is routed to the specified weighted destinations.
+
+The destinations appear as references in `backendRefs`.
+In the case that a request should be routed to the service it would otherwise
+have been routed to, `backendRefs` can be omitted.
+
+```yaml
+- backendRefs:
+  - weight: 90
+    kind: MeshServiceSubset
+    name: backend
+    tags:
+      version: v2
+  - weight: 10
+    kind: MeshServiceSubset
+    name: backend
+    tags:
+      version: v2
+  ...
+```
+
+Note that it's not a valid configuration to have a `requestRedirect` and
+`backendRefs` specified or both `requestRedirect` and `urlRewrite`.
+
+Note that in Gateway API, filters are also available on a _per-backend_ basis (`HTTPBackendRef.filters`).
+We leave this option open for our policy in the future.
+
+#### Load balancing
+
+At the moment `TrafficRoute` can be used to configure load balancing behavior at
+the top, rule-independent level. This MADR proposes **removing** load balancing from
+the route policies and instead later creating a separate policy to handle load
+balancing configuration.
+
+The main motivator for this is that load balancing is done on an Envoy cluster.
+Including load balancing options on a match would force creation of a new cluster for that
+set of `backendRefs` which has negative consequences including but not limited to
+metrics handling and a potentially large number of clusters.
+
+#### Merging behavior
+
+With `targetRef` resources, the ultimate configuration of a data plane proxy
+is always a result of all policies targeting that proxy depending on their
+specificity.
+
+> :question: Is the current merge behavior what we need for routing policy? What surprises are there? See the next sections
+
+This proposal addresses one issue with a "naive" application of `targetRef`
+structure to a routing policy.
+
+##### Surprising behavior of top level `default`
+
+Let's look at two basic theoretical route policies where we include `default` at
+the top level of a `to` rule:
+
+```yaml
+metadata:
+ name: owner
+spec:
+ targetRef:
+  kind: Mesh
+ to:
+  - targetRef:
+     kind: MeshService
+     name: backend
+    default:
+     rules:
+      - matches:
```

Comment:
I was reading Gateway API, and I noticed `matches` is a list https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRouteRule. For example in a single rule we can match 2 different paths:
```yaml
rules:
  - matches:
      - path:
          prefix: /foo
      - path:
          prefix: /bar
```
Did we intentionally create `matches` as a single element?

Reply:
no, it should be a list. the examples are incorrect.